lib/internal/url.js

```diff
@@ -618,97 +644,183 @@ class URL {
   }
 
   set href(value) {
-    const valid = updateUrl(this.#context.href, updateActions.kHref, `${value}`, this.#onParseComplete);
-    if (!valid) { throw ERR_INVALID_URL(`${value}`); }
+    const href = bindingUrl.update(this.#context.href, updateActions.kHref, `${value}`);
+    if (!href) { throw ERR_INVALID_URL(`${value}`); }
+    this.#updateContext(href);
   }
 
   // readonly
   get origin() {
-    return this.#context.origin;
+    const protocol = StringPrototypeSlice(this.#context.href, 0, this.#context.protocol_end);
+
+    if (URLContext.SpecialProtocols.has(protocol)) {
```

Comment:
Couldn't this be made faster?

ada::url_aggregator, like ada::url, has a type value. The type is ada::scheme::type which is one of those...

```
enum type : uint8_t {
  HTTP = 0,
  NOT_SPECIAL = 1,
  HTTPS = 2,
  WS = 3,
  FTP = 4,
  WSS = 5,
  FILE = 6
};
```

So there is no need to construct a view and check whether it belongs to the set...

```
static SpecialProtocols = new SafeSet(['http:', 'https:', 'ws:', 'wss:', 'ftp:', 'file:']);
```

You can just grab `type` and check whether it is different from 1. And to check whether it is a file, you just need to compare it with 6. It is going to be free compared to the string comparisons, hashing and stuff that the current code is doing.

Reply:
You're right. I'm updating the PR. Thanks for the review @lemire